docs/proposals/memory-management.md

```diff
@@ -0,0 +1,165 @@
+
+8. All burstable containers that exceed their request are equally likely to be killed. Ideally, the job that exceeds the most should be killed. 
+  
+##### How to address these limitations?
+
+We can improve memory management by making use of [kernel memcg features](https://www.kernel.org/doc/Documentation/cgroups/memory.txt), and by modifying the current system behavior. Some of these improvements are required in the short-term. The rest are either complex or require significant changes to the system.
+
+###### What can be done right away?
+
+1. Cleanup the entire container, whenever a OOM kill is detected in a container. Users should be able to override the default behavior for containers that have an in-built process manager that can handle OOM kills.
+
+2. Set memory soft limits on Burstable containers to their request. This should result in "Burstable" containers being pushed to their requested memory limit under system memory pressure. 
+
+3. Restrict resource usage of system processes. Since the clusters are heterogeneous, the kubelet will have to come up with a limit for system processes dynamically, and apply those limits. 
+
+4. Kubelet should expose `available` memory in addition to `capacity` and `usage`. The scheduler can then schedule containers based on `available` instead of `capacity`, which is the case as of now.
+To begin with `available` can be the difference between `capacity` and the total memory assigned to all system processes.
+
+5. Schedule `Best-effort` containers based on worst-fit scheduling, that takes into account the usage. This will reduce the over-commitment of memory to some extent.
+
+
+###### What else can be done?
+
+*Note: The features mentioned here are not recommended to be added to the system until the changes mentioned above are completed, and we have concrete evidence for further improving memory management in the system.*
+
+1. Avoid overcommiting memory assigned to `Guaranteed` tasks by imposing limits on `Best-effort` and `Burstable` containers. By organizing memory cgroups hierarchically, `burstable` can be limited to use only the memory that has not been allocated for `Guaranteed` tasks. `Best-effort` containers can use the memory allocated to `Burstable` containers.
+
+2. Provide pod level Out of Memory handling: If any of the containers in a pod dies, the entire pod will be cleaned up.
+
+3. Kubelet attempts to reduce the likelihood of system memory pressure:
+
+   a. Kubelet can subscribe to memory pressure notifications from the kernel, and possibly evict low priority containers (or)
+
+   b. Kubelet can set thresholds based notifications on the root cgroup and evict containers that are most suitable candidates based on the QoS policy and current usage. Kubelet can try to limit total memory usage to under 90% of the memory capacity.
+   
+   This change requires defining the concept of priority and restart SLA for containers and pods in the system.
```

Comment:
Once resource metrics are plumbed into the scheduler, the scheduler can avoid assigning more pods to a node under distress. Also, see my note in the doc about surfacing `availability` from the node.